pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -572,16 +576,69 @@ func matchFailureDomains(gMS *machineapi.MachineSet, rMS machineapi.MachineSet,
 	rMS_providerconfig, err := cpms.NewProviderConfigFromMachineSpec(logr, rSpec, infrastructure)
 	if err != nil {
 		logger.WithError(err).Errorf("unable to parse remote MachineSet %v provider config", rMS.Name)
-		return false, err
+		return false, nil, err
 	}
 	gMS_providerconfig, err := cpms.NewProviderConfigFromMachineSpec(logr, gSpec, infrastructure)
 	if err != nil {
 		logger.WithError(err).Errorf("unable to parse generated MachineSet %v provider config", gMS.Name)
-		return false, err
+		return false, nil, err
 	}
 	fd1 := rMS_providerconfig.ExtractFailureDomain()
 	fd2 := gMS_providerconfig.ExtractFailureDomain()
-	return fd1.Equal(fd2), nil
+
+	if rMS_providerconfig.Type() == "AWS" {
+		subnets := []*ec2.Subnet{}
+		awsactuator := actuator.(*AWSActuator)
+
+		if aws.StringValue(fd1.AWS().Subnet.ID) == "" {
+			subnets, err = getSubnetIDByFilter(fd1.AWS().Subnet, *awsactuator)
+			if err != nil {
+				return false, nil, err
+			}
+		}
+		subnets, err = awsactuator.filterPublicSubnets(subnets, pool)
+		if err != nil {
+			return false, nil, err
+		}
+
+		if aws.StringValue(subnets[0].AvailabilityZone) != fd2.AWS().Placement.AvailabilityZone {
+			return false, subnets, nil
+		} else {
+			return true, subnets, nil
+		}
+	}
+
+	return fd1.Equal(fd2), nil, nil
+}
+
+func getSubnetIDByFilter(subnet *v1.AWSResourceReference, actuator AWSActuator) ([]*ec2.Subnet, error) {
+	for _, filter := range *subnet.Filters {
+		if filter.Name == "tag:Name" { // FIXME use all filters or just this one?
+			req := &ec2.DescribeSubnetsInput{
+				Filters: []*ec2.Filter{
+					{
+						Name:   aws.String("tag:Name"),
+						Values: aws.StringSlice(filter.Values),
+					},
+				},
+			}
+			resp, err := actuator.awsClient.DescribeSubnets(req)
+			if err != nil {
+				return nil, err
+			}
+			if len(resp.Subnets) == 0 {
+				return nil, fmt.Errorf("no subnets found for filter %v", filter)
+			}
+			// compare subnets to generated machineset
+			// FIXME do I compare all subnets in list? How does this work
+			if len(resp.Subnets) > 1 {
+				return nil, fmt.Errorf("multiple subnets found for filter %v", filter)
+			}
```

Comment:
Yeah, gosh, I really don't know. Might be worth asking someone in MAPI-land if we're expecting exactly one subnet, and should error if we find otherwise. Is the public/private thing still in play?